### ISSUE TITLE
At what time(s) is the template function invoked?

### ISSUE THREAD

**Issue by marcoms** [skatejs/skatejs#170]
I've recently discovered skatejs and I've been studying its API in preparation for using it in a project. One thing that I don't quite understand is the `template` function. Does it get automatically on a state change, or is it just ran on element creation?

The example given in README.md seems to suggest that it is called once:

    function shadowDomTemplate (shadowHtml) {
      return function (element) {
          element.createShadowRoot().innerHTML = shadowHtml;
      };
    }

...since multiple `createShadowRoot` calls will create a new shadow root if another exists, but it would be nice to have confirmation of `template`'s behaviour.

PS. If `template` is only called on creation, then what would be the difference between putting template code in `created`/`attached`, vs in `template`?

**Comment by treshugart** [comment]
The `template` callback gets executed only once before the `created` callback is invoked. If you put template code in attached it will get called whenever the element is attached to the DOM. This means if you add it to the DOM, it will get templated. If you then remove it and add it back into the DOM, it will get called again. This is most probably undesirable behaviour, but nothing is stopping you from doing it.

Let me know if that helps.

**Comment by treshugart** [comment]
There's this in case you haven't seen it: http://skatejs.github.io/best-practices.html. The section on lifecycle callbacks might help.

**Comment by marcoms** [comment]
Thanks, that was very helpful indeed!
